### DISCUSSION TITLE
Graphical Canvas fingerprinting (demo sites & planning)

### DISCUSSION THREAD

**Discussion by daijro** [daijro/camoufox#41] [Fingerprinting]
**This discussion is dedicated to researching & implementing a safe, undetectable way to spoof the HTML Canvas pixel data in Camoufox without causing leaks.**

## Testing

Here is a test site I put together demonstrating the issue:

https://camoufox.com/tests/canvas

**End Goals**:
- Randomization should not be detectable
- Hashes should change in new sessions
- Hashes should not change between retests/regenerations of the same image

## Randomization

Firefox currently has built-in functionality to spoof the canvas fingerprint by randomizing the image buffer before it is returned:

https://searchfox.org/mozilla-central/source/dom/canvas/CanvasRenderingContext2D.cpp#2078

However, WAFs easily detect this by creating a random canvas with known results, reading it back, and checking the differences (see [this paper](https://antoinevastel.com/tracking/2018/07/01/eval-canvasdef.html), the demo website [here](https://arkenfox.github.io/TZP/tests/canvasnoise.html) by the Arkenfox team).

## Strategy

A more subtle and less detectable approach could involve introducing randomization earlier in the rendering pipeline, specifically at the operation level.

By randomizing individual rendering steps (anti-aliasing, sub-pixel rendering), or the algorithms used for drawing shapes and text (for example, by randomly offsetting the position similar to how Camoufox already [handles font fingerprinting](https://camoufox.com/fingerprint/fonts/#font-metrics)), we can avoid creating obvious image artifacts.

**Comment by ThomasAlbrt** [comment]
God I wish I could help you.

But the truth is I'm not good enough to even know where to start.

**Comment by vlrevolution** [comment]
How's this going? Is this relevant perhaps?
https://github.com/daijro/camoufox/issues/121

**Comment by Ghxst** [comment]
Vendors have a list of known GPU Vendor + Renderer combinations for a collection of canvas ops, they will always produce the same result on the same hardware, you cannot introduce variance in any part of the rendering pipeline that changes the hash without being penalized, I've tested multiple GPUs with dozens of drivers and the biggest variance you see between drivers for the same GPU + renderer combination is less than a handful so this is a very strong signal for them.

**Comment by jeygeethan-rocketapex** [comment]
@daijro I have a usecase where I want to disable randomization of the canvas for a particular user. If I start a session for a particular user, the canvas fingerprint should be the same when i start the browser again for the same user, say 1 day after. Is this possible?

**Comment by semandar** [comment]
Camoufox only has canvas randomization now? How do I make canvas constant at least say within one session? Is there any ways to make non-unique canvas?

**Reply by daijro** [reply]
Camoufox randomizes Skia's anti-aliasing algorithm, not the full Canvas. It also uses the same offsets per session to prevent leaks. This passes all known Canvas noise tests used in production, but there are certain low-entropy tests that it can fail (that are not commonly implemented due to false positives on some systems).

I'll add an option to disable it in the next release of the Python library.

**Reply by semandar** [reply]
I mean right now canvas literally changes every 5 seconds. That is not a good disguise because 99 percent of the time it indicates a disguise. A unique canvas is a red flag. Canvas has to be non-unique to get lost in the crowd.

**Comment by nttrung9x** [comment]
Hi @daijro 

I am also testing the code on skia library, but I don't know how to switch firefox profile to use skia, I have set in about:config, but firefox still uses webrender, can you guide me how to switch back to skia?

Currently, if I use webrender, I can only fake the canvas data part as randoms, and the Hardware.canvas part in browserscan does not change, do you know what parameters browserscan checks to change Hardware.canvas?

Thank you!

**Comment by ghost** [comment]
@daijro, 

Is it possible to turn off patched skia canvas rendering? I tried to switch to other renderers, but no luck. Still getting unique canvas fp, which I' like to avoid. From my pov it's always better to have non-unique canvas fingerprint. Otherwise browser getting detected as anti-detect. 

Hope to hearing from you.

**Reply by vlrevolution** [reply]
> @daijro,
> 
> Is it possible to turn off patched skia canvas rendering? I tried to switch to other renderers, but no luck. Still getting unique canvas fp, which I' like to avoid. From my pov it's always better to have non-unique canvas fingerprint. Otherwise browser getting detected as anti-detect.
> 
> Hope to hearing from you.

This is true, we need persistent sessions or it beats the purpose of any other efforts
